### PR TITLE
Static Logging

### PR DESCRIPTION
**Pull request by ajohnstonTE**
Updates Gemini and all Gemini modules to use slf4j for logging, and adds three new quick-start modules that provide log4j2/logback bindings for slf4j plus sample log4j2/logback config files that produce nearly the exact same logging as before (and nearly the exact same logging as each other).

- The exact same logging as before _could_ have been achieved, but I decided to completely ditch the use of component codes in favor of classes to allow for package-level filtering.

### CODE REVIEW

File: gemini/src/main/java/com/techempower/gemini/log/package-info.java [hunk 1]
```diff
@@ -1,31 +0,0 @@
...
- * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
- * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
- * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
- * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
- *******************************************************************************/
-
-/**
- * Gemini-specific log components.
- */
-package com.techempower.gemini.log;
```
[the remark this answers is not in the document]
Comment (ajohnstonTE) [reply gemini/src/main/java/com/techempower/gemini/log/package-info.java:31]: Sounds good

File: gemini/src/main/java/com/techempower/gemini/log/GeminiComponentLog.java [hunk 2]
```diff
@@ -83,14 +69,13 @@ public void setStringValue(String s)
    * with each log message.
    */
   private static final ThreadLocalContextInfo CONTEXT_INFO = new ThreadLocalContextInfo();
-  
+
   /**
    * Constructor.  Generally not invoked directly, but rather through
    * GeminiApplication.getLog().
```
[the remark this answers is not in the document]
Comment (ajohnstonTE) [reply gemini/src/main/java/com/techempower/gemini/log/GeminiComponentLog.java:75]: Never. At this point this class isn't even a log, it's purely used for the `CONTEXT_INFO` ThreadLocal. It's what provides the `{sessionNumber} - {requestNumber} - {userId}` in the logging during requests, which is inserted into the logs automatically using the classes in the various logging modules like `gemini-logback`'s [RequestPatternConverter](https://github.com/TechEmpower/gemini/blob/cf7605f91b07005b01dc35c79283db6395f32cba/gemini-logback/src/main/java/com/techempower/gemini/log/RequestPatternConverter.java) or [the class of the same name](https://github.com/TechEmpower/gemini/blob/cf7605f91b07005b01dc35c79283db6395f32cba/gemini-log4j2/src/main/java/com/techempower/gemini/log/RequestPatternConverter.java) in `gemini-log4j2`. I kept the name of the `GeminiComponentLog`class the same for legacy purposes, but I doubt anyone uses this class directly so I'll probably just rename it. I'll also update this comment.

File: gemini/src/main/java/com/techempower/gemini/GeminiApplication.java [hunk 3]
```diff
@@ -1324,12 +1307,8 @@ protected void runShutdownTasks()
         }
         catch (Exception exc)
         {
-          // In the event a shutdown tasks throws an exception, we will report
-          // the exception to stderr (because the log may be closed) and then
-          // proceed to the next shutdown task.
```
[the remark this answers is not in the document]
Comment (ajohnstonTE) [reply gemini/src/main/java/com/techempower/gemini/GeminiApplication.java:1329]: At least in logback, the flushing of logs on app shutdown should be handled for you automatically. http://logback.qos.ch/manual/configuration.html#webShutdownHook

[the remark this answers is not in the document]
Comment (ajohnstonTE) [reply gemini/src/main/java/com/techempower/gemini/GeminiApplication.java:1329]: Also, on http://logback.qos.ch/manual/appenders.html:
> By default, each log event is immediately flushed to the underlying output stream. This default approach is safer in the sense that logging events are not lost in case your application exits without properly closing appenders. However, for significantly increased logging throughput, you may want to set the `immediateFlush` property to false.

File: gemini/src/main/java/com/techempower/gemini/cluster/jms/CacheMessageManager.java [hunk 4]
```diff
@@ -484,19 +472,10 @@ else if (group instanceof EntityGroup)
...
               // it as a CacheGroup, only an EntityGroup, which means we have
               // nothing to update here.
-              log.debug("Received 'cache object expired' for an " +
-                  "EntityGroup, so not cached. Ignoring:{}", cacheMessage);
             }
-            else
-            {
-              log.info("Receiving 'cache object expired' but group id is " +
-                  "invalid:{}, group: {}", cacheMessage, group);
-              application.getNotifier().addNotification(
```
Comment (msmith-techempower) [inline gemini/src/main/java/com/techempower/gemini/cluster/jms/CacheMessageManager.java]: Why was this removed? I think some applications use this.

Comment (ajohnstonTE) [reply gemini/src/main/java/com/techempower/gemini/cluster/jms/CacheMessageManager.java]: Hm. Weird. Seems like git's auto-resolve thing with the merge conflict felt like removing this. Will add it back and re-review what it removed.

Comment (ajohnstonTE) [reply gemini/src/main/java/com/techempower/gemini/cluster/jms/CacheMessageManager.java]: Actually, this wasn't part of the conflict. Apparently I removed this. What?

Comment (ajohnstonTE) [reply gemini/src/main/java/com/techempower/gemini/cluster/jms/CacheMessageManager.java]: Oh, wait, never mind, I didn't remove it. I was just confused by the line history I was looking at. Keith removed this line recently: https://github.com/TechEmpower/gemini/commit/9064e0aa99d85d8e07a98452f7bad6bbc0f6c007#diff-f840e62d66613d77a961f9408873d9a3L135-L140
